modules/ding_user/ding_user.module

```diff
@@ -451,58 +451,66 @@ function ding_user_default_authname($name) {
   // characters, leaving enough room for our suffix.
   $hash = hash('sha512', drupal_get_private_key() . $name, TRUE);
 
-  // The string length of the hash have been changed in the line below and we
-  // need to update the accounts with new hash values, hence this function call.
-  _ding_user_update_old_accounts($hash);
-  return _password_base64_encode($hash, drupal_strlen($hash)) . '@ding_user';
+  // Hence the strlen().
+  $authname = _password_base64_encode($hash, strlen($hash)) . '@ding_user';
+
+  _ding_user_update_authname($authname);
+
+  return $authname;
 }
 
 /**
- * This method updates the user authname hash.
+ * Updates authname for existing users.
  *
- * It will become deprecated after a while.
+ * Brings compatibility of user accounts across various php versions.
+ * Issue caused by drupal_strlen() -> mb_strlen(), whose
+ * behavior was changed somewhat in php 5.4+
  *
- * @param string $hash
- *   Hash used as provider users authname.
+ * @see http://php.net/releases/NEWS_5_4_0_beta2.txt
+ * - "Ill-formed UTF-8 check for security enhancements."
  */
-function _ding_user_update_old_accounts($hash) {
-  // Create authname hash using old workflow (the difference is in strlen VS
-  // drupal_strlen).
-  $auth_name_old = _password_base64_encode($hash, strlen($hash)) . '@ding_user';
+function _ding_user_update_authname($authname) {
+  // We'll check for first 32 chars to get a match.
+  // Hope _password_base64_encode() gives unique results.
+  $authname_chunk = substr($authname, 0, 32);
 
-  // Try to find the users uid based on the old hash value (length of the old
-  // hash should be 96).
-  $uid = db_select('authmap', 'a')
+  // Pick the user id that has almost the authname value that we seek,
+  // which is still different from what it should be.
+  $query = db_select('authmap', 'a')
     ->fields('a', array('uid'))
-    ->condition('authname', $auth_name_old)
+    ->condition('authname', $authname_chunk . '%', 'LIKE');
```

Comment:
This makes me edgy. I would prefer to add another condition that only selects authnames where strlen(authname) < strlen(proper_authname), so it'll only consider the "short" authnames. Should be just another ->condition().

Having a collision in the first 32 chars (which is 24 bytes of the hash) is unlikely, but sorting out good authnames feels better.